Sockets/BackEndMonitoringSockets/src/CClientConnectionHandler.cpp

```diff
@@ -1,105 +1,145 @@
+		result = HandleRequestEvent(socket_fd, EEventType::REQUEST_PROCESS_DATA);
 		break;
 	case EEventType::CLOSE_EVENT:
-		if (HandleExitEvent(socket_fd, message) == true)
-		{
-			result = false;
-		}
-		break;
+		HandleExitEvent(socket_fd); // If we can exit result equal false
+		return false;
 	default:
 		result = false;
 		break;
 	}
 
+	if (result)
+	{
+		CLOG_DEBUG_WITH_PARAMS("Result is true, start HandleResponseEvent, for the socket", socket_fd);
+		result = HandleResponseEvent(socket_fd, message);
+	}
+
 	return result;
 }
 
 bool CClientConnectionHandler::HandleRequestEvent(const int socket_fd, 
-	EEventType type, std::string& message)
+	EEventType type)
 {
-	std::string request_str;
-	switch (type) {
-	case EEventType::REQUEST_ALL_DATA:
-		request_str = "ALL_DATA";
-		break;
-	case EEventType::REQUEST_DISK_DATA:
-		request_str = "DISK_DATA";
-		break;
-	case EEventType::REQUEST_PROCESS_DATA:
-		request_str = "PROCESS_DATA";
-		break;
-	default:
-		return false;
-	}
-	if (m_p_client_stream->Send(socket_fd, request_str) == false)
+	std::string request_str = ConvertRequestToString(type);
+	CLOG_DEBUG_WITH_PARAMS("Convert event type to the string", request_str);
+	if (SendRequestToServer(socket_fd, request_str))
 	{
-		return false;
+		return true;
 	}
-	m_can_make_request = false;
-	return HandleResponseEvent(socket_fd, message);
+	return false;
 }
 
 bool CClientConnectionHandler::HandleResponseEvent(const int socket_fd, 
 	std::string& message)
 {
+	bool result = m_p_client_stream->Receive(socket_fd, message);
+	CLOG_DEBUG_WITH_PARAMS("We send request to the server, result", result);
+	if (result)
+	{
+		// If we get the wrong request from the server
+		if (message.empty() || message == "-1")
+		{
+			CLOG_ERROR_WITH_PARAMS("Response from the server", message);
+			return false;
+		}
+		else if (message == "Request lost") // If we lost request
+		{
+			CLOG_ERROR_WITH_PARAMS("Response from the server", message);
+			result = HandleLostRequestEvent(socket_fd, message);
+		}
+	}
 
-	message = m_p_client_stream->Receive(socket_fd);
-	if (message == "-1")
+	if(result)
 	{
-		return false;
+		return HandleDataReceivedEvent(socket_fd);       
 	}
 
-	return HandleDataReceivedEvent(socket_fd);
+	return false;
+}
+
+bool CClientConnectionHandler::HandleExitEvent(const int socket_fd)
+{
+	return SendRequestToServer(socket_fd, "Exit");
+}
+
+bool CClientConnectionHandler::HandleDataReceivedEvent(const int socket_fd)
+{
+	CLOG_DEBUG("Handle data received event");
+	return m_p_client_stream->Send(socket_fd, "DATA RECEIVED");
 }
 
-bool CClientConnectionHandler::HandleExitEvent(const int socket_fd, 
+bool CClientConnectionHandler::HandleLostRequestEvent(const int socket_fd, 
 	std::string& message)
 {
-	m_p_client_stream->Send(socket_fd, "Exit");
-	int count_iteration = 0;
-	while (true)
+	const int max_trials = 5;
+	int count_trials = 0;
+	std::string request_str = ConvertRequestToString(m_current_request);
+
+	if(request_str == "ERROR")
```

Comment:
Use compare function instead of equal operator

Reply:
Fixed

Reply:
But I really don't understand why this:
request_str.compare("ERROR") == c_equal;
 is better than this:
 request_str == "ERROR";
Using it in this way can be confusing:
!request_str.compare("ERROR");
and simple zero we shouldn't use because it is a magic number.

Reply:
Because '==' could be dangerous, in some custom/3rdParty strings it could compare pointers to the data instead of the data. It is very easy to change std::string to another string class on any project. Most real projects prefer to use custom strings.
Also if you will write 
"request_str.compare("ERROR") != 0;"
nobody will blame you that zero is the magic number(because everybody knows what 0 means in compare method). Note: It's related only to string compare.

Reply:
Now I understand, thank you!